stan/math/fwd/scal/fun/log_diff_exp.hpp

```diff
@@ -23,16 +26,25 @@ inline fvar<T> log_diff_exp(const fvar<T>& x1, const fvar<T>& x2) {
 template <typename T1, typename T2>
 inline fvar<T2> log_diff_exp(const T1& x1, const fvar<T2>& x2) {
   using std::exp;
-  if (x1 <= x2.val_)
+  if (x1 <= x2.val_) {
+    if (x1 < INFTY && x1 == x2.val_)
+      return fvar<T2>(NEGATIVE_INFTY, x2.d_ * NEGATIVE_INFTY);
```

Comment:
What's the logic of multiplying by negative infinity here? Is it so that the derivative will be positive or negative infinity depending on the sign of ```x2.d_```?

Reply:
Yes. If x1==x2 then the result is negative infinite but if x2 is even a tiny bit smaller the result is finite so the derivative must be infinite and opposite in sign to the derivative of x2.
Also, if x2.d_ == 0 then it's ambiguous and NaN is the only possible derivative. Take for example `log_diff_exp(0, -exp(k*x))`. When x is very negative this is approximately `log(1-(1-exp(k*x)) == k*x` so the derivative should be k at `x==NEGATIVE_INFTY`. But exp "forgets" k and the autodiff sees only `log_diff_exp(0, fvar(0, 0))`.

Reply:
That makes sense, but if the purpose is to determine the sign of INFTY, wouldn't it be more efficient to use the [copysign function](https://en.cppreference.com/w/cpp/numeric/math/copysign) rather than a multiplication?

Reply:
I guess so, I'm not too familiar with C++. But that still leaves the NaN case and if we really care about efficiency it might be best to just ignore the derivatives at infinity, I can't really imagine those ever being useful in practice.

Reply:
This is coming up in other PRs and my understanding is that @syclik wants the boundaries to behave properly.

Multiplication is probably faster than anything involving a branch as long as it doesn't involve autodiff.

Reply:
That all sounds good to me